### PR TITLE
feat(connection pool hooks): Add afterPoolAcquire and beforePoolRelease hooks

### PR DESCRIPTION
**Pull request by b3n01t-o**
Provide `afterPoolAcquire` and `beforePoolRelease` hooks.
It addresses: #8966 

This is not my use case thought. 
I need to be able to interact with the connection right before and right after it is being used to run a query or open a transaction on the SQL server.

### Pull Request check-list

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](../CONTRIBUTING.md)?

### Description of change

Provide `afterPoolAcquire` and `beforePoolRelease` hooks.

### CODE REVIEW

File: lib/dialects/abstract/connection-manager.js [hunk 1]
```diff
@@ -273,6 +273,10 @@ class ConnectionManager {
     return promise.then(() => {
       return this.pool.acquire(options.priority, options.type, options.useMaster)
         .then(mayBeConnection => this._determineConnection(mayBeConnection))
+        .then(definitlyConnection => {
+          this.sequelize.runHooks('afterPoolAcquire', definitlyConnection);
```
Comment (sushantdhiman) [inline lib/dialects/abstract/connection-manager.js]: Hooks can be async

```js
this.sequelize.runHooks('afterPoolAcquire', connection)
  .return(connection)
  .tap(() => { debug('connection acquired'); });
```

File: lib/dialects/abstract/connection-manager.js [hunk 2]
```diff
@@ -273,6 +273,10 @@ class ConnectionManager {
     return promise.then(() => {
       return this.pool.acquire(options.priority, options.type, options.useMaster)
         .then(mayBeConnection => this._determineConnection(mayBeConnection))
+        .then(definitlyConnection => {
```
Comment (sushantdhiman) [inline lib/dialects/abstract/connection-manager.js]: Just call it `connection`

File: lib/dialects/abstract/connection-manager.js [hunk 3]
```diff
@@ -285,6 +289,7 @@ class ConnectionManager {
    * @return {Promise}
    */
   releaseConnection(connection) {
+    this.sequelize.runHooks('beforePoolRelease', connection);
```
Comment (sushantdhiman) [inline lib/dialects/abstract/connection-manager.js]: Hooks can be async

File: test/integration/hooks/hooks.test.js [hunk 4]
```diff
@@ -134,6 +134,34 @@ describe(Support.getTestDialectTeaser('Hooks'), () => {
       });
     });
 
+    describe('After acquire connection hook', () => {
```
Comment (sushantdhiman) [inline test/integration/hooks/hooks.test.js]: Add unit tests for this change instead https://github.com/sequelize/sequelize/blob/master/test/unit/connection-manager.test.js

**Comment by b3n01t-o** [comment]
@sushantdhiman thank you for your review. I Hope I took care your comments properly. 

The AppVeyor tests fail because of:

`FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory`

**Comment by sushantdhiman** [comment]
> CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory

Not good, let's not add this feature for now.

> I need to be able to interact with the connection right before and right after it is being used to run a query or open a transaction on the SQL server.

You can add beforeQuery / afterQuery hook if you want, that can cover your case

**Pull request closed without merging**